saasherder/changelog.py

```diff
@@ -0,0 +1,158 @@
+        useful = [s for s in changed if s.get('hash') and old(s)]
+
+        # Now changed services can be new ones or change in hash
+        return [{
+            'name': service['name'],
+            'url': service['url'],
+            'new': service.get('hash', 'master'),  # NOTE: null here is silly
+            'old': old(service)
+        } for service in useful]
+
+    @staticmethod
+    def run(cmd):
+        return subprocess.run(cmd, stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE,
+                              shell=True, check=True)
+
+    def worktree(self, service_name):
+        """Get git worktree path for a service"""
+
+        return self.workspace + service_name
+
+    def _clone(self, service):
+        """Private; use checkout instead"""
+
+        cmd = "git clone {} {}".format(service['url'],
+                                       self.worktree(service['name']))
+        self.run(cmd)
+
+    def _fetch(self, service):
+        """Private; use checkout instead"""
+
+        worktree = self.worktree(service['name'])
+        cmd = "cd {} && git fetch ".format(worktree)
+
+        return self.run(cmd)
+
+    def checkout(self, service, version):
+        """Get a service checked out at specified version"""
+
+        if os.path.exists(self.worktree(service['name'])):
+            self._fetch(service)
+        else:
+            self._clone(service)
+
+        worktree = self.worktree(service['name'])
+        cmd = "cd {} && git checkout {} ".format(worktree, version)
+
+        return self.run(cmd)
+
+    def log(self, service, start, end):
+        """Get the changelog for one service"""
+
+        worktree = self.worktree(service['name'])
+        cmd = "cd {} && git log --oneline {}..{} ".format(worktree, end, start)
+
+        return self.run(cmd).stdout.decode('utf-8')
+
+    @staticmethod
+    def markdown(changelog, start, end):
+        """Pretty print the change log
+
+        The incoming object is a list of `(name, messages)` tuples.
+        """
+
+        header = "# Changelog \n\n## From version {} to {}".format(start, end)
+
+        body = ["## {} \n{} \n".format(name, messages)
+                for (name, messages) in changelog]
+
+        return header + ''.join(body)
+
+    def main(self, context, start, end):
+
+        # Checkout to previous version and get services
+        self.run("git checkout {}".format(end))
+        previous = self.services(context)
+
+        self.run("git checkout {}".format(start))
+        now = self.services(context)
+
+        # Go back to where we were
+        self.run("git checkout master")
+
+        changed = self.diff(previous, now)
+
+        # Fetch all the services that changed
+        for service in changed:
+            self.checkout(service, service['new'])
+
+        # Get the changelog for each service as a `(name, [messages])` tuple
+        changelog = [(service['name'], self.log(service, start, end)) for
```

Comment:
`start` and `end` here are not actually the relevant hash for what changed in the services. Thus it is actually not showing the true diff.

Should use the start/end from the service info.

Reply:
@maxandersen +1. My bad :/ 👍